### PR TITLE
Custom braille tables (#3304)

### LINKED ISSUE #3304: Custom braille tables
**Reported by falinn.onda on 2013-06-25 11:49**
Following the discussion in
http://community.nvda-project.org/ticket/3284#comment:5

 I think it is a good idea that NVDA will enable users to customize tables,
 if they know the table format (by referring to louise docs).

In response to jteh
" \* Coming up with an elegant user interface for this is tricky.
- We then have to document and explain how it works.
- It encourages users to copy files into the NVDA program directory. In
  general, users should never touch the NVDA program directory, only their
  config directory."

jteh, i agree with both comments.
I did not mean that NVDA will add a graphical user interface or that NVDA will have full documentation about customizing tables.
I do not think that is necessary since there is already a very good table definition format defined and explained in detail by louise documentation.
I think other users can tweak tables to their needs like i was able to after reading louise documentation.

the only thing i am suggesting is to have several table files which the user will be able to edit. I agree that it is bad practice to edit the default standard tables in louise/tables and that is exactly why i think it is a good idea to have few table files in userConfig\brailletables (or similar). 
The only thing i need from NVDA is to have an option in the braille setting input/output tables combo boxes which lets me choose to use those custom files i wrote.
It is the user responsibility to edit the file outside NVDA in a text editor of his choice  (or get a file someone else wrote).).

For example if we can put other1_custom_user_table.ctb other2_custom_user_table.ctb and other3_custom_user_table.ctb
in userConfig\brailleTables. (i'm suggesting file names starting with o because there is no such tables that start with o right now which will make it easy to switch to those combo )entries by pressing o ).

I am aware that not everyone will understand how to edit tables but most advanced users will benefit from this option and maybe give the tables to other users in their community or publish online like addons. 

In NVDA documentation we can only add a chapter titled "Customizing braille tables":
Advanced users who wish to use custom braille tables can edit the tables files x, y, z in the folder w. For a definition of the file format see luise documentation link.
Note that users who only wish to add symbols to ecisting tables can use the include command.

Here we can have a small example file that includes english braille table and adds additional letter and punctuation as example. 

What do you think?
**Blocking #4996**

### LINKED ISSUE #9863: Braille: The fall-back input table does not match the default input table.
### Steps to reproduce:

Manually alter the configuration key ["braille"]["inputTable"] to an incorrect value.

### Actual behavior:

Braille input falls back to using `en-us-comp8.ctb`

### Expected behavior:

Braille input should fall back to using `en-ueb-g1.ctb`, as it is the default for a new configuration.

#### NVDA installed/portable/running from source:
Any
#### Windows version:
N/A
#### Name and version of other software in use when reproducing the issue:
N/A
#### Does the issue still occur after restarting your PC?
Yes

### PR DESCRIPTION
**Pull request by JulienCochuyt**
### Link to issue number:

Fixes #3304.
Fixes #9863.
Supersedes PR #9864.
Addresses https://github.com/nvaccess/nvda/issues/505#issuecomment-311438115
Provides a workaround for #3470

### Summary of the issue:

In NVDA, there is no easy and reliable way for an add-on to provide a new braille table.
For an experienced users wishing to do so there are two options:
 - Alter manually an existing table in `louis/tables`.
  The new table still has its original name in the settings GUI.
  This change is lost upon NVDA updates.
 - Set the absolute path to the table file in the configuration in lieu of the usual file name.
  The settings GUI shows an empty entry for this one.
  Forces to manually alter `nvda.ini`.
  Forces to copy in the same directory the whole dependency chain of the new table plus `braille-patterns.cti`. This is because liblouis default table resolver only looks for tables in a single directory, See https://github.com/nvaccess/nvda/pull/5489#discussion_r46647804

### Description of how this pull request fixes the issue:

- Add a new `brailleTables` optional directory in both the user scratchpad directory and the add-on directory structure.
- Implement a custom liblouis table resolver which looks up first in these directories, then in `louis/tables`.
- Support reading tables metadata from an optional `brailleTables` section of the add-on manifest or from a `.ini` file with the same format found in the `brailleTables` subdirectory of the scratchpad directory, allowing a user to provide a display names and set output/input/contracted capabilities with no code.
- Enforce the existing fallback mechanism to ensure there still is braille output if the configured table cannot be found eg. because an add-on or the scratchpad directory was disabled. This now applies both to the main configuration and individual profiles and also covers braille input.

As is already the case in the default liblouis table resolver, table file names are treated as an identifier.
That is, if a user eg. provides in its `scratchpad/brailleTables` directory a new version of a table which is imported by other tables, all of those are impacted.
Only one version of a given table file name can be resolved. In that concern, the user scratchpad directory takes precedence over add-ons, which take precedence over the tables provided with NVDA.
It also mean that the manifest file can alter metadata of a table file provided elsewhere.
Eg. adding a manifest to the `scratchpad/brailleTables` directory is enough to workaround #3470.

Manifest file example as found in the developer guide:

```
[brailleTables]
[[fr-bfu-tabmod-comp8.utb]]
displayName = French (unified) 8 dot computer braille - Tab mod
contracted = False
output = True
input = True
[[no-no-generic.ctb]]
displayName = Norwegian unofficial computer braille table
contracted = False
output = True
input = True
```

Example add-on demonstrating the new functionality (please rename .zip to .nvda-addon): 
[pr10712-customBrailleTables-demo-2021.05.08.zip](https://github.com/nvaccess/nvda/files/6446631/pr10712-customBrailleTables-demo-2021.05.08.zip)

### Testing strategy:

From a source copy:
 - added a new table via an add-on
 - added a new table via the scratchpad directory
 - replaced an existing table via an add-on
 - replaced an existing table via the scratchpad directory
 - altered existing table meta-data via an add-on
 - altered existing table meta-data via the scratchpad directory
 - ensured proper fallback when the user selected a new table which was provided by a later disabled add-on
 - ensured proper fallback when the user selected a new table which was provided by a later removed add-on
 - ensured proper fallback when the user selected a new table which was provided by the later disabled scratchpad directory

### Known issues with pull request:

The current automatic translation of older tables names to newer tables names as of a liblouis update can lead to unexpected behavior if there is collision, but I do not think I am the one to take the decision when to drop this backward-compatibility piece.

## Change log entry:

Section: Changes for developers
Add-on can now provide custom braille tables.

### Code Review Checklist:

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [ ] System (end to end) tests.
- [x] Manual tests.
- [x] User Documentation.
- [x] Change log entry.
- [ ] Context sensitive help for GUI changes.

### CODE REVIEW

File: source/brailleTables.py
Comment (leonardder) [inline source/brailleTables.py]: I guess this can be removed?

File: source/brailleTables.py
Comment (leonardder) [inline source/brailleTables.py]: Shouldn't this be logged?

File: source/brailleTables.py
Comment (leonardder) [inline source/brailleTables.py]: I guess you're avoiding the bare exception check here?

```suggestion
				except Exception:
```

File: source/brailleTables.py
Comment (leonardder) [inline source/brailleTables.py]: This looks a bit odd. Could you use the f literal here?

File: source/brailleTables.py
Comment (leonardder) [inline source/brailleTables.py]: I do not understand this line.

File: source/brailleTables.py
Comment (leonardder) [inline source/brailleTables.py]: This is no longer true.

File: source/brailleTables.py
Comment (leonardder) [inline source/brailleTables.py]: I like this style. Do you think it makes sense to log if the isdir check doesn't pass?

File: source/brailleTables.py
Comment (leonardder) [inline source/brailleTables.py]: Have you considered making this a property on CustomTablesManifest?

**Review by LeonarddeR — COMMENTED** [review]
Thanks for this huge and nice work! I think I like the switch from json to configobj.

Now I wonder, have you considered storing table data in the add-ons manifest itself? I think this has a major benefit, namely that the add-on manifest parser has support for language inheritance out of the box. This means that coming up with the right translated name for the table is the responsibility of the potential add-on translator, not of the initial author.

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py]: Oups, damn dev log... I'm removing this one right away!
It is probably worth adding a debug log of the whole `tablesDirs` list passed at `louisHelper` initialization though.

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py]: `tablesDirs[:0] = customDirs` is a "slice-replace".
A slower equivalent would be:
```
for dir_ in reversed(customDirs):
    tablesDirs.insert(0, dir_)
```
That is, it prepends `tableDirs` with the content of `customDirs` while ensuring an eventual external reference to `tablesDirs` remains valid.
Another way of writing it, but invalidating an external reference, would be:
`tablesDirs = customDirs + tablesDirs`

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py]: You are right, this one has been removed as of your pre-PR comments of the Python 2 version.

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py]: I would guess no.
Having a file named "brailleTables" is damn unlikely, and no one would reasonably expect its content to be interpreted.
The check could even be removed here and such a file would still not break the resolver.

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py]: Missed this one when porting from Python 2. Thanks!

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py]: Why not.
I've turned the check into `try..except FileNotFoundError` for added robustness.

**Comment by JulienCochuyt** [comment]
> Thanks for this huge and nice work! I think I like the switch from json to configobj.

You are very welcome. Thanks for you interest.

> Now I wonder, have you considered storing table data in the add-ons manifest itself? I think this has a major benefit, namely that the add-on manifest parser has support for language inheritance out of the box. This means that coming up with the right translated name for the table is the responsibility of the potential add-on translator, not of the initial author.

The add-on manifest could indeed safely contain a "brailleTables" section with the needed configuration. `brailleTables._getCustomTableDisplayName` even handles the case where "displayName" is a key rather than a section - which fits this use case just fine.

My concern, though, is that ConfigObj does not support formally optional sections.
See DiffSK/configObj#131
This means that the "brailleTables" section could not be specified in the add-on manifest specification unless every single add-on manifest provides it. I guess neither of these alternatives is fine.

Anyway, an add-on can safely add a custom table during its initialization and have its display name translated as usual:
Eg: `brailleTables.addTable("somefile.utb", , _("Some fancy name"), output=True, input=True, contracted=False)`

Writing this I realize that `brailleTables._loadCustomTablesConfig` should have its leading underscore removed as the following is also a valid use for an add-on:

```
brailleTables.loadCustomTablesConfig({
    "file1.utb": {
        "displayName": _("Name of first table"),
        "output": True,
        "input": True,
        "contracted": True
    },
    "file2.utb: {
        "displayName": _("Name of second table"),
        "output": True,
        "input": True,
        "contracted": True
    }
})
```

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py]: As you prefer.
If I'm not wrong, both versions lead to the exact same result.
I felt that the `noqa` comment was more honest in admitting it circumvents the rule for the sake of simplicity and robustness, but that's a matter of style.

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py]: Dev log removed as of ea43238.
Debug log added as of af30589.

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py]: Comment added as of deb3836.

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py]: Actually, no.
The manifest is only one way to supply the config.
Another is to pass a dictionary to `brailleTables.loadCustomTablesConfig` (made public as of b87f5b22, see https://github.com/nvaccess/nvda/pull/10172#issuecomment-529537805)

**Comment by leonardder** [comment]
@feerrenrut, could your share your thoughts about what you think would be best to specify metadata for custom braille tables?

**Comment by DrSooom** [comment]
See also: https://github.com/liblouis/liblouis/wiki/Table-discovery-based-on-table-metadata

**Comment by leonardder** [comment]
@JulienCochuyt wrote:

> This means that the "brailleTables" section could not be specified in the add-on manifest specification unless every single add-on manifest provides it.

Is this true? Is there anything holding you back from providing a [brailleTables] section and a [[__many__]] section in there and providing a manifest that doesn't have this section? I gess the validator would just add the empty section just as it adds defaults for optional fields, such as readme.

**Comment by JulienCochuyt** [comment]
> @JulienCochuyt wrote:
> 
> > This means that the "brailleTables" section could not be specified in the add-on manifest specification unless every single add-on manifest provides it.
> 
> Is this true? Is there anything holding you back from providing a [brailleTables] section and a [[**many**]] section in there and providing a manifest that doesn't have this section? I gess the validator would just add the empty section just as it adds defaults for optional fields, such as readme.

It at least is how I understood DiffSK/configobj#131 back then: You can't have a specified optional section. Unspecified sections are supported, but once they are specified, they become mandatory.
Nevertheless, I might try to play a bit more with ConfigObj to try ensure that point.

But anyhow, I doubt bundling a hundreds of tables in an add-on would ever be a reasonable use case, and keep in mind that registering the tables via simple code already allows translators to follow their usual workflow.

**Comment by leonardder** [comment]
> ... keep in mind that registering the tables via simple code already allows translators to follow their usual workflow.

Problem with this is that there are multiple pathways to do exactly the same. I'd rather have only one public method of bundling tables in add-ons.

**Comment by JulienCochuyt** [comment]
> > ... keep in mind that registering the tables via simple code already allows translators to follow their usual workflow.
> 
> Problem with this is that there are multiple pathways to do exactly the same. I'd rather have only one public method of bundling tables in add-ons.

Actually, the support for metadata `.ini` files was initially merely added to easily add custom tables in the scratchpad directory with no code - extending my personal use case of wanting my own modified fr-bfu-comp8 available when disabling all add-ons. Felt then like providing non-developer end users with as much icing on the cake as possible.
Thought it also provided an easy way to allow building of a minimal add-on with no code that would eg. add the requested no-no-comp table. I never considered an add-on author would really want to provide dozens of tables with localized names - but still made sure to support it.

We're in Python, there isn't a single way to do things, even if a compatriot of yours is told to be the one able to distinguish which is the right one :)

If I understand your point of view, it's hard to advertise the metadata `.ini` files way as it does not fit with the standard translation process. Though, I can remove it altogether if you are more comfortable with it.

**Comment by leonardder** [comment]
> Actually, the support for metadata `.ini` files was initially merely added to easily add custom tables in the scratchpad directory with no code - extending my personal use case of wanting my own modified fr-bfu-comp8 available when disabling all add-ons. Felt then like providing non-developer end users with as much icing on the cake as possible.

Mm yes, this is an interesting one. This requires a method that doesn't involve add-on manifests.

> We're in Python, there isn't a single way to do things, even if a compatriot of yours is told to be the one able to distinguish which is the right one :)

This is true. However, I'd rather advertise one public method to do it, leaving it up to an add-on author to use another method that might be perfectly valid, but isn't the advised one. For example, if we decide that manifests is the way to go, add-on authors can still use the registration method, but they are encouraged to use the manifest method.

> If I understand your point of view, it's hard to advertise the metadata `.ini` files way as it does not fit with the standard translation process. 

Yes, this is true. However, I think it is up to either @feerrenrut or @michaelDCurran to crack the nut here.

> Though, I can remove it altogether if you are more comfortable with it.

I'd say lets leave it as is until we have some more opinions about this.

**Comment by leonardder** [comment]
Could either @michaelDCurran and @feerrenrut elaborate on what NV access' standpoint is with regard to what implementation to use to define custom tables? See https://github.com/nvaccess/nvda/pull/10172#issuecomment-529537805

**Comment by feerrenrut** [comment]
Apologies for the lack of input from us on this PR. It looks like this is going to take some time to get up to speed on. I'll set aside some time later this week to look at this in more detail.

**Comment by JulienCochuyt** [comment]
@feerrenrut wrote:
> Apologies for the lack of input from us on this PR. It looks like this is going to take some time to get up to speed on. I'll set aside some time later this week to look at this in more detail.

No problem. I just hope it still works as well as when I thoroughly tested it back then :)

**Comment by feerrenrut** [comment]
Ok, there is a lot to take in on this PR. Is it possible to reduce the scope so we can handle different concerns one at a time. I haven't looked carefully at the code yet, but perhaps it's possible to decide on translations / or addon support mechanisms until later. They could maybe be handled in a subsequent PR, and just start with supporting scratchpad and "name is how it is".

I agree this that addon's should be able to provide this eventually, but planning for add-ons creates a lot of unknowns. It would be really useful to see some examples of how addons are expected to use this. While there may be multiple possible ways of doing things, it would be nice if we had one "recommended way". 

I don't want to create a situation where we wish to make an alteration to this design but don't want to break compatibility for add-ons.  Perhaps call it a "design in beta", warning add-on devs that the design may change without updating the API version. It would be nice if add-ons had to acknowledge this through the manifest?

I'm going to come back to this PR

File: source/brailleTables.py
Comment (feerrenrut) [inline source/brailleTables.py:19]: Rather than duplicate the definition for the default, you could look up the default value from the config spec: `config.conf.getConfigValidation(("braille", "cursorBlinkRate")).kwargs["max"]`

**Comment by lukaszgo1** [comment]
@feerrenrut I'm afraid I have to disagree with the order of priorities here. The main use case should be add-ons and the scratchpad dir in a subsequent PR. If someone has enough knowledge to use scratchpad he/she is probably capable of creating an add-on for his/her personal usage. Also at the moment this cannot be done for add-ons in a non-hacky way   at all, and having this for add-ons would benefit bigger user base. To give some real life example back in 2017 I've created Polish translation of LaTeX-access braille table, however due to the differences in the literary and computer braille table it was not possible to settle on character representation that would work in both. Having an ability  to set custom braille table would eliminate a lot of problems for me back then.

**Comment by leonardder** [comment]
I agree with @feerrenrut that we must be absolutely sure about the way 
add-ons have to provide custom tables is the right way from the start. 
However, I think it is safe to declare something subject to change when 
add-ons yet are allowed to use it, without breaking compatibility, that 
is, compatibility is allowed to break as long as we explicitly stated 
that the API wasn't stable.

**Comment by JulienCochuyt** [comment]
@feerrenrut wrote:
> Ok, there is a lot to take in on this PR. Is it possible to reduce the scope so we can handle different concerns one at a time. I haven't looked carefully at the code yet, but perhaps it's possible to decide on translations / or addon support mechanisms until later. They could maybe be handled in a subsequent PR, and just start with supporting scratchpad and "name is how it is".

I can for sure understand you concerns. Mine are that I already spent quite a lot of time on this subject.
Part of the design is justified only as it aims at supporting easy translation and add-ons support.
Splitting is of course always possible, but I'm concerned there might be much less motivation to examine the second package once the first one gets merged, which would defeat the goal of this PR altogether.
Users that are able to put the right files in their scratchpad directory and won't bother having to deal with technical filenames instead of friendly labels are most likely to be already able to find out a quick-and-dirty way to have braille presented or input as they like.
This full-fledged solution aims more at aiding novices to feel at home right away, eventually by installing a small addon built in literally five minutes by a friendly volunteer.

> I agree this that addon's should be able to provide this eventually, but planning for add-ons creates a lot of unknowns. It would be really useful to see some examples of how addons are expected to use this. While there may be multiple possible ways of doing things, it would be nice if we had one "recommended way".

I can for sure provide examples and technical guidelines to be discussed.

> I don't want to create a situation where we wish to make an alteration to this design but don't want to break compatibility for add-ons. Perhaps call it a "design in beta", warning add-on devs that the design may change without updating the API version. It would be nice if add-ons had to acknowledge this through the manifest?

Flagging the API as beta by any means seems very reasonable.
The custom table(s) manifest could also be versioned, or it could stay as simple and short as can be for its first version and require a version indicator if it ever has to evolve.

> I'm going to come back to this PR

Thanks.

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py:19]: Makes sense. Thanks for pointing it out.

Comment (JulienCochuyt) [reply source/brailleTables.py]: Done: e25872c
Thanks for teaching me how to retrieve these default values.

**Comment by JulienCochuyt** [comment]
> I can for sure provide examples and technical guidelines to be discussed.

What form would you prefer for these?
Shall I simply post here or do you prefer we open a wiki section somewhere?

**Comment by JulienCochuyt** [comment]
After quite some time, I realize my argument in https://github.com/nvaccess/nvda/pull/10172#issuecomment-529537805 was wrong regarding optional section support in ConfigObj: It doesn't matter as long as NVDA does not write back the add-on manifest. And even if it does, it would just add an empty section. No big deal.

So, as proposed by @leonardder in https://github.com/nvaccess/nvda/pull/10172#pullrequestreview-285504609 the custom tables meta-data is now provided directly in the add-on manifest, eventually translated by the usual means.
If this design is approved, I'll also submit to the addonTemplate repo my changes in the `sconstruct` to handle it smoothly.
The same meta-data can also be provided in the scratchpad in the exact same format.
For simplicity, I dropped altogether the ability to provide display names in different locales in the scratchpad.
The other parts of the scratchpad do not support custom localization anyway.

Another notable change is the use of a `ChainMap` instead of the former dictionary to store both standard and custom tables.
This allows, once the whole thing is squashed onto master, to preserve the history of the individual `addTable` calls for the standard tables.

All these change make for a much smaller patch, and theoretically slightly better performance.

I've attached to the description of this PR an add-on demonstrating the new functionality.

Finally, sorry @DrSooom I realize I never answered your https://github.com/nvaccess/nvda/pull/10172#issuecomment-533593685: Table discovery through their embedded meta data is not really feasible/wishable here because:
 - An add-on or the scratchpad might provide a replacement table which is referenced by another one, but not intended to be available for selection by the user.
 - Embedded meta-data concerning contraction vary depending on the target locale
 - NVDA already ships its standard tables with display names that differ from those embedded

Anyway, thanks for pointing to this interesting article.

**Review by seanbudd — DISMISSED** [review]
Thanks @JulienCochuyt, I've given this a preliminary review but I would still wait on any thoughts from @feerrenrut, and a further review from myself. There's some minor merge conflicts in the docs that need to be fixed as well.

File: source/braille.py
Comment (JulienCochuyt) [reply source/braille.py:2083]: I agree it reads better, but it also is not equivalent.
In the original snippet as written for #7712 the renamed driver name is also set to config.
If done all in one line, you either loose readability or end up in always writing to the config even when not necessary,

File: source/braille.py
Comment (JulienCochuyt) [reply source/braille.py:2097]: Same as above, this is not equivalent.

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py]: Done as of 4a70af0cc.

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py:103]: It also has been reverted to ease blaming. There of course are means to exclude revisions when blaming (and we'll need it if eventually linting the whole source tree), but I've thought it would be much easier this way.
Please note that in the current proposal, `terminate` does restore the module in the state it was before calling `initialize`.
Both methods really do `initialize` / `terminate` only the handling of custom tables.

@feerrenrut, @leonardder may we have your thoughts here before rewriting this part yet another time?

File: source/brailleTables.py
Comment (JulienCochuyt) [reply source/brailleTables.py:486]: Of course, but I guess it'll depend on the consensus around https://github.com/nvaccess/nvda/pull/10172#discussion_r637675471

**Comment by JulienCochuyt** [comment]
@seanbudd wrote:
> There's some minor merge conflicts in the docs that need to be fixed as well.

Done as of f3be0f9f7.

File: source/brailleTables.py
Comment (leonardder) [reply source/brailleTables.py:103]: I prefer leaving this as is and doing this in a follow up pr.

Comment (seanbudd) [reply source/brailleTables.py:103]: It appears that the consensus is to leave as is.

File: source/braille.py
Comment (seanbudd) [inline source/braille.py:2349]: when resolving this merge conflict, I kept this line where it currently is in master. However, it might need to get moved to `handlePostConfigProfileSwitch`?

Comment (feerrenrut) [reply source/braille.py:2349]: Why do you think it should be there? I don't think it needs to re-initialize after every config update?

**Review by feerrenrut — COMMENTED** [review]
I'm going to need to come back to this next week.

**Comment by burmancomp** [comment]
How is it with this pull request? I think this would be useful feature.

**Comment by leonardder** [comment]
@JulienCochuyt any chance you can bring this forward? If not, I think you did some very valuable work here and I'm happy to take this one. I'll do that if I'm not hearing back from you within a month or so.

**Comment by amirmahdifard** [comment]
time flys: one day someone created this, then people reviewed it such as @feerrenrut and now, he is no longer working in nvda project, so many things has been changed: where is the life taking us, i'm becoming sad whenever i look at nvda's old pull requests. Rest in peace. Sorry, i couldn't stop my self to share my emotions.

**Comment by Adriani90** [comment]
@LeonarddeR I think @JulienCochuyt is no longer that active on Github, at least I didn't see any updates from him since many years. Are you still thinking about taking this up?

**Comment by Adriani90** [comment]
Just a note, in my view unlike #3304 suggests, I would rather prefer an approach similar to speech dictionaries where user's dictionaries are stored in the user config and where users can edit the dictionaries in a GUI. . But I don't know if this is technically feasible. Nevertheless, this PR definitely would point things in the right direction at least.

**Comment by seanbudd** [comment]
Hi - I'm going to mark this as a draft. I know that the reason this has been abandoned due to a delayed review from NV Access, however I do not see it in our near capacity to fix up the merge conflicts and prepare this for another review.
Given this is now the only item in our ready, unblocked, PR queue, I think we can be confident that this will be reviewed in a timely manner in future if further work is done.

**Comment by LeonarddeR** [comment]
I'm definitely taking this, since I might have a use case for myself in the near future. That's a pretty good stick behind the door if I may say so😉

**Comment by LeonarddeR** [comment]
Superseeded by #16208

**Comment by CyrilleB79** [comment]
Since it has been superseded, I remove the "Abandoned" label.

@seanbudd I hope NV Access approves such triage actions; it should allow to filter Abandoned PRs for somewone wanting to take it over.

**Comment by seanbudd** [comment]
Thanks @CyrilleB79

**Pull request closed without merging**
